tsl/src/chunk_copy.c

```diff
@@ -0,0 +1,581 @@
+/*
+ * This file and its contents are licensed under the Timescale License.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-TIMESCALE for a copy of the license.
+ */
+#include <postgres.h>
+#include <foreign/foreign.h>
+#include <catalog/pg_foreign_server.h>
+#include <catalog/pg_foreign_table.h>
+#include <catalog/dependency.h>
+#include <catalog/namespace.h>
+#include <access/htup_details.h>
+#include <access/xact.h>
+#include <nodes/makefuncs.h>
+#include <utils/acl.h>
+#include <utils/builtins.h>
+#include <utils/syscache.h>
+#include <utils/inval.h>
+#include <utils/tuplestore.h>
+#include <utils/palloc.h>
+#include <utils/memutils.h>
+#include <utils/snapmgr.h>
+#include <executor/executor.h>
+#include <parser/parse_func.h>
+#include <funcapi.h>
+#include <miscadmin.h>
+#include <fmgr.h>
+
+#if USE_ASSERT_CHECKING
+#include <funcapi.h>
+#endif
+
+#include <compat.h>
+#include <chunk_data_node.h>
+#include <extension.h>
+#include <errors.h>
+#include <error_utils.h>
+#include <hypertable_cache.h>
+
+#include "chunk.h"
+#include "chunk_api.h"
+#include "chunk_copy.h"
+#include "data_node.h"
+#include "remote/dist_commands.h"
+#include "dist_util.h"
+
+#define CCS_INIT "init"
+#define CCS_CREATE_EMPTY_CHUNK "create_empty_chunk"
+#define CCS_CREATE_PUBLICATION "create_publication"
+#define CCS_CREATE_REPLICATION_SLOT "create_replication_slot"
+#define CCS_CREATE_SUBSCRIPTION "create_subscription"
+#define CCS_SYNC_START "sync_start"
+#define CCS_SYNC "sync"
+#define CCS_DROP_PUBLICATION "drop_publication"
+#define CCS_DROP_SUBSCRIPTION "drop_subscription"
+#define CCS_ATTACH_CHUNK "attach_chunk"
+#define CCS_DELETE_CHUNK "delete_chunk"
+
+typedef struct ChunkCopyStage ChunkCopyStage;
+typedef struct ChunkCopy ChunkCopy;
+
+typedef void (*chunk_copy_stage_func)(ChunkCopy *);
+
+struct ChunkCopyStage
+{
+	const char *name;
+	chunk_copy_stage_func function;
+	/* todo: abort function */
+};
+
+/* To track a chunk move or copy activity */
+struct ChunkCopy
```

Comment:
I think we should keep the FormData struct in catalog.h and instead follow the established pattern of emdedding the FormData in a larger struct  like follows:


```
typedef struct FormData_chunk_copy_activity
    {
      
    	int32 id;
    	NameData operation_id;
    	int32 backend_pid;
    	text *completed_stage;
    	TimestampTz time_start;
    	int32 chunk_id;
    	NameData source_node_name;
    	NameData dest_node_name;
    	bool delete_on_src_node;
};

struct ChunkCopy
{
     struct FormData_chunk_copy_activity fd;
     ChunkCopyStage *stage;
     Chunk *chunk;
     ForeignServer *src_server;
     ForeignServer *dst_server;
     MemoryContext mctx;
}
```

I would also consider making the `completed_stage` a NameData type instead of text (I don't think we need more than 63  chars for the stage names), since that would make the FormData struct fixed size. That in turn allows using GETSTRUCT on the tuple instead of `heap_deform`.